### ISSUE TITLE
fit method hangs after recent updates

### ISSUE THREAD

**Issue by paxtonfitzpatrick** [lmcinnes/umap#367]
I can't figure out why this is happening, but after some recent changes to `master` branch, the `.fit` method hangs to the point I have to kill python to avoid it crashing my computer.

The problem doesn't exist as of the most recent 0.4 release candidate (fc59aa78f14b10e5eafc19b18fc135de1c5382ea), but does with the most recent version of `master` (1a3892610643ca0ba72a0c559488f47e43bbd7cf).

It appears to be coming from a super slow list comprehension `scipy` runs internally to transpose a sparse matrix, called from `umap` [here](https://github.com/lmcinnes/umap/blob/1a3892610643ca0ba72a0c559488f47e43bbd7cf/umap/umap_.py#L1696).  It doesn't seem like a bug in `scipy` though, since my version is the same for both the working and broken examples (`1.2.1`).

<details>
<summary>Working version:</summary>

```
import numpy as np
from time import time
from umap import UMAP

to_reduce = np.random.uniform(0, 1, (5000, 25))
reducer = UMAP(random_state=0, metric='euclidean', verbose=True)

t1 = time()
reducer.fit(to_reduce)
t2 = time()
print(t2 - t1)

--------------------------------------------------------
UMAP(a=None, angular_rp_forest=False, b=None,
     force_approximation_algorithm=False, init='spectral', learning_rate=1.0,
     local_connectivity=1.0, low_memory=False, metric='euclidean',
     metric_kwds=None, min_dist=0.1, n_components=2, n_epochs=None,
     n_neighbors=15, negative_sample_rate=5, output_metric='euclidean',
     output_metric_kwds=None, random_state=0, repulsion_strength=1.0,
     set_op_mix_ratio=1.0, spread=1.0, target_metric='categorical',
     target_metric_kwds=None, target_n_neighbors=-1, target_weight=0.5,
     transform_queue_size=4.0, transform_seed=42, unique=False, verbose=True)
Construct fuzzy simplicial set
Mon Mar  2 14:57:26 2020 Finding Nearest Neighbors
Mon Mar  2 14:57:26 2020 Building RP forest with 9 trees
Mon Mar  2 14:57:26 2020 NN descent for 12 iterations
	 0  /  12
	 1  /  12
	 2  /  12
	 3  /  12
	 4  /  12
	 5  /  12
	 6  /  12
	 7  /  12
	 8  /  12
	 9  /  12
	 10  /  12
	 11  /  12
Mon Mar  2 14:57:33 2020 Finished Nearest Neighbor Search
Mon Mar  2 14:57:35 2020 Construct embedding
	completed  0  /  500 epochs
	completed  50  /  500 epochs
	completed  100  /  500 epochs
	completed  150  /  500 epochs
	completed  200  /  500 epochs
	completed  250  /  500 epochs
	completed  300  /  500 epochs
	completed  350  /  500 epochs
	completed  400  /  500 epochs
	completed  450  /  500 epochs
Mon Mar  2 14:57:47 2020 Finished embedding
21.280228853225708
```
</details>

<details>
<summary>Broken version:</summary>

```
import numpy as np
from time import time
from umap import UMAP

to_reduce = np.random.uniform(0, 1, (5000, 25))
reducer = UMAP(random_state=0, metric='euclidean', verbose=True)

t1 = time()
reducer.fit(to_reduce)
t2 = time()
print(t2 - t1)

--------------------------------------------------------
UMAP(a=None, angular_rp_forest=False, b=None,
     force_approximation_algorithm=False, init='spectral', learning_rate=1.0,
     local_connectivity=1.0, low_memory=False, metric='euclidean',
     metric_kwds=None, min_dist=0.1, n_components=2, n_epochs=None,
     n_neighbors=15, negative_sample_rate=5, output_metric='euclidean',
     output_metric_kwds=None, random_state=0, repulsion_strength=1.0,
     set_op_mix_ratio=1.0, spread=1.0, target_metric='categorical',
     target_metric_kwds=None, target_n_neighbors=-1, target_weight=0.5,
     transform_queue_size=4.0, transform_seed=42, unique=False, verbose=True)
Construct fuzzy simplicial set
Mon Mar  2 14:52:09 2020 Finding Nearest Neighbors
Mon Mar  2 14:52:09 2020 Building RP forest with 9 trees
Mon Mar  2 14:52:10 2020 NN descent for 12 iterations
	 0  /  12
	 1  /  12
	 2  /  12
	 3  /  12
	 4  /  12
	 5  /  12
	 6  /  12
	 7  /  12
	 8  /  12
	 9  /  12
	 10  /  12
	 11  /  12
Mon Mar  2 14:52:17 2020 Finished Nearest Neighbor Search
---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-3-e2b9ca9bf5a6> in <module>()
      1 t1 = time()
----> 2 reducer.fit(to_reduce)
      3 t2 = time()
      4 print(t2 - t1)

/opt/conda/lib/python3.6/site-packages/umap/umap_.py in fit(self, X, y)
   1694                 self._search_graph.data = _data
   1695                 self._search_graph = self._search_graph.maximum(
-> 1696                     self._search_graph.transpose()
   1697                 ).tocsr()
   1698 

/opt/conda/lib/python3.6/site-packages/scipy/sparse/lil.py in transpose(self, axes, copy)
    456 
    457     def transpose(self, axes=None, copy=False):
--> 458         return self.tocsr(copy=copy).transpose(axes=axes, copy=False).tolil(copy=False)
    459 
    460     transpose.__doc__ = spmatrix.transpose.__doc__

/opt/conda/lib/python3.6/site-packages/scipy/sparse/lil.py in tocsr(self, copy)
    473 
    474         indptr = np.cumsum([0] + lst, dtype=idx_dtype)
--> 475         indices = np.array([x for y in self.rows for x in y], dtype=idx_dtype)
    476         data = np.array([x for y in self.data for x in y], dtype=self.dtype)
    477 

/opt/conda/lib/python3.6/site-packages/scipy/sparse/lil.py in <listcomp>(.0)
    473 
    474         indptr = np.cumsum([0] + lst, dtype=idx_dtype)
--> 475         indices = np.array([x for y in self.rows for x in y], dtype=idx_dtype)
    476         data = np.array([x for y in self.data for x in y], dtype=self.dtype)
    477 

KeyboardInterrupt: 
```
</details>

Note: this only happens when the data isn't "small" and the distance matrix doesn't get pre-computed

**Comment by lmcinnes** [comment]
Thanks for this -- this is certainly something that is non-obvious from the point of view of code changes. In particular I am surprised that the pre-release worked and master didn't since I am not sure how this code path managed to change. I'll try to take a look at it soon. Thanks again for the detailed report; it helps a lot in tracking these sorts of errors/issues down.

**Comment by paxtonfitzpatrick** [comment]
No problem, thanks @lmcinnes! I have #336 nearly fixed, just can't test it fully due to this

**Comment by paxtonfitzpatrick** [comment]
@lmcinnes figured it out.  The issue came from this commit: cde7c6dc8c69bb4cde30c6241924fa22e3c2039d, in
```
_rows = []
 _data = []
for i in range(self._knn_indices.shape[0]):
    _rows.append(self._knn_indices[self._knn_indices >= 0])
     _data.append(np.ones(_rows[-1].shape[0], dtype=np.int8))
self._search_graph.rows = _rows
self._search_graph.data = _data
```

`_rows` was getting the entire `self._knn_indices` matrix appended to it every time, so `self._search_graph.data` ended up containing 5,000x lists instead of int8's, with 375 _million_ nonzero items, instead of 75,000.

Something else odd happened with this change as well -- previously, `self._search_graph.rows` contained the indices of the nearest neighbors and `self._search_graph.data` contained an array of 1's for neighbors with non-zero distances, and 0's elsewhere.  Now, `self._search_graph.rows` contains the indices of the nearest neighbors (with the -1 issue fixed) and `self._search_graph.data` contains an array of _all_ 1's, for _all_ nearest neighbors, **including those with distances of 0**.  Was that change intended? 
If so, this should work:
```
_rows = []
 _data = []
for i in self._knn_indices:
    _non_neg = i[i >= 0]
    _rows.append(_non_neg)
    _data.append(np.ones(_non_neg.shape[0], dtype=np.int8))
self._search_graph.rows = _rows
self._search_graph.data = _data
```

Otherwise, this would work:
```
_rows = []
 _data = []
for i in self._knn_indices:
    _non_neg = i[i >= 0]
    _rows.append(_non_neg)
    _data.append((_non_neg > 0).astype(np.int8))
self._search_graph.rows = _rows
self._search_graph.data = _data
```

Let me know which to go with, and I'll add this fix when I submit a PR for #366 🙂

**Comment by lmcinnes** [comment]
Sorry -- I thought I had commented on this already, but apparently I managed not the hit the button to have it sent/posted. Ideally we should be including distance 0 points as searching over those will improve things (presuming they somehow have different neighbors). I believe that means we want option 1.
